planner/core/exhaust_physical_plans.go

```diff
@@ -142,6 +142,24 @@ func (p *LogicalJoin) getMergeJoin(prop *property.PhysicalProperty) []PhysicalPl
 		mergeJoin.OtherConditions = p.moveEqualToOtherConditions(offsets)
 		mergeJoin.initCompareFuncs()
 		if reqProps, ok := mergeJoin.tryToGetChildReqProp(prop); ok {
+			// Adjust expected count for children nodes.
+			if prop.ExpectedCnt < p.stats.RowCount {
+				expCntScale := prop.ExpectedCnt / p.stats.RowCount
+				reqProps[0].ExpectedCnt = p.children[0].statsInfo().RowCount * expCntScale
+				reqProps[1].ExpectedCnt = p.children[1].statsInfo().RowCount * expCntScale
+				if len(p.LeftJoinKeys) > 0 {
+					leftStats, rightStats := p.children[0].statsInfo(), p.children[1].statsInfo()
+					outerKeyCardinality := getCardinality(p.LeftJoinKeys, p.children[0].Schema(), leftStats)
+					innerKeyCardinality := getCardinality(p.RightJoinKeys, p.children[1].Schema(), rightStats)
+					outIdx := 0
+					if p.JoinType == RightOuterJoin {
+						outIdx = 1
+						outerKeyCardinality, innerKeyCardinality = innerKeyCardinality, outerKeyCardinality
+					}
+					cardinalityRatio := innerKeyCardinality / outerKeyCardinality
+					reqProps[1-outIdx].ExpectedCnt = reqProps[outIdx].ExpectedCnt * cardinalityRatio
```

Comment:
Should we take `inner.RowCount` into account?
In this line, `inner.RowCount` is ignored.
What does this formula mean?

Reply:
Oh, sorry my mistake. I revisited this strategy:
> How about this simple strategy?
Since we only need ExpectedCnt rows from outer, the estimated cardinality of outer's result is ExpectedCnt * outer.Cardinality / outer.rowCount.
And the cardinality scale of inner's result is inner.rowCount / inner.Cardinality.
So in order to get all keys which appear in outer's result from inner, we have to fetch (ExpectedCnt * outer.Cardinality / outer.rowCount) * (inner.rowCount / inner.Cardinality) rows from inner, which should be inner's ExpectedCnt?

and noticed that the formula `(ExpectedCnt * outer.Cardinality / outer.rowCount) * (inner.rowCount / inner.Cardinality)` didn't consider the ratio of outer.Cardinality and inner.Cardinality. If we take that into consideration, the formula would be like:
```
(ExpectedCnt * outer.Cardinality / outer.rowCount) * (inner.Cardinality / outer.Cardinality) * (inner.rowCount / inner.Cardinality)
```
which is `(ExpectedCnt / outer.rowCount) * inner.rowCount` actually.